src/cpu/flexcpu/simple_dataflow_thread.cc

```diff
@@ -0,0 +1,890 @@
+                "Decoded instruction (seq %d) - %#x : %s\n",
+                inst_ptr->seqNum(),
+                pc.instAddr(),
+                decode_result->disassemble(pc.instAddr()).c_str());
+
+        if (decode_result->isMacroop()) {
+            curMacroOp = decode_result;
+
+            inst_ptr->staticInst(decode_result);
+            inst_ptr->squash();
+
+            advanceInst();
+        } else {
+            inst_ptr->staticInst(decode_result);
+
+            issueInstruction(inst);
+        }
+
+    } else { // If we still need to fetch more MachInsts.
+        fetchOffset += sizeof(TheISA::MachInst);
+        onBeginFetch(inst);
+    }
+}
+
+void
+SDCPUThread::onPCTranslated(weak_ptr<InflightInst> inst, Fault fault,
+                            const RequestPtr& req)
+{
+    const shared_ptr<InflightInst> inst_ptr = inst.lock();
+    if (!inst_ptr || inst_ptr->isSquashed()) {
+        // No need to do anything for an instruction that has been squashed.
+        return;
+    }
+
+    const TheISA::PCState pc = inst_ptr->pcState();
+
+    DPRINTF(SDCPUThreadEvent, "onPCTranslated(pc: %#x)\n",
+                              pc.instAddr());
+
+
+    if (fault != NoFault) {
+        markFault(inst_ptr, fault);
+
+        return;
+    }
+
+    DPRINTF(SDCPUThreadEvent, "Received PC translation (va: %#x, pa: %#x)\n",
+            req->getVaddr(), req->getPaddr());
+
+    auto callback = [this, inst](TheISA::MachInst data) {
+        onInstDataFetched(inst, data);
+    };
+
+    // No changes to the request before asking the CPU to handle it if there is
+    // not a fault. Let the CPU generate the packet.
+    if (!_cpuPtr->requestInstruction(req, callback)) {
+        panic("The CPU rejected my fetch request and I haven't been programmed"
+              " to know how to continue!!!");
+    }
+}
+
+void
+SDCPUThread::populateDependencies(shared_ptr<InflightInst> inst_ptr)
+{
+    // NOTE: instead of implementing dependency detection all in this one
+    //       function, maybe it's better to define a dependency provider
+    //       interface, to make these rules swappable?
+
+    // Temporary virtual dependency rule to simulate in-order CPU mode.
+    // if (inflightInsts.size() > 1) {
+    //     const shared_ptr<InflightInst> last_inst =
+    //                                             *(++inflightInsts.rbegin());
+
+    //     if (!(last_inst->isComplete() || last_inst->isSquashed())
+    //      && !last_inst->staticInst()->isMacroop())
+    //         inst->addDependency(last_inst);
+    // }
+
+    const StaticInstPtr static_inst = inst_ptr->staticInst();
+
+    // BEGIN Sequential Consistency
+
+    // TODO IMPORTANT this still doesn't make it safe to speculate on memory
+    //      instructions, since if we produce a fault and need to squash, we
+    //      still need a way to inform the cpu to undo stores being sent to
+    //      memory.
+
+    if (static_inst->isMemRef() || static_inst->isMemBarrier()) {
```

Comment:
Add a comment here that you're searching for the youngest memory instruction that has been issued and creating a dependency.